third_party/intel/lib/TritonAnnotateModule/TritonAnnotateModule.cpp

```diff
@@ -0,0 +1,70 @@
+#include "intel/include/Analysis/DPAS.h"
+#include "intel/include/Dialect/TritonIntelGPU/IR/Dialect.h"
+#include "intel/include/TritonAnnotateModule/Passes.h"
+
+namespace mlir::triton::gpu::intel {
+#define GEN_PASS_DEF_TRITONANNOTATEMODULE
+#include "intel/include/TritonAnnotateModule/Passes.h.inc"
+} // namespace mlir::triton::gpu::intel
+
+using namespace mlir;
+using namespace mlir::triton::gpu;
+using DPASAnalysis = intel::DPASAnalysis;
+
+namespace {
+
+struct TritonAnnotateModule
+    : intel::impl::TritonAnnotateModuleBase<TritonAnnotateModule> {
+  using Base::Base;
+
+  void runOnOperation() final {
+    ModuleOp mod = getOperation();
+    if (target.getValue().empty()) {
+      mod.emitError("Expecting target specification");
+      return signalPassFailure();
+    }
+
+    Builder builder(mod);
+    mod->setAttr(intel::TritonIntelGPUDialect::getTargetAttrName(),
+                 builder.getStringAttr(target.getValue()));
+
+    // FIXME: Use SYCL runtime to query supported OpenCL extensions, instead
+    // of checking driver version.
+    if (isLTS)
+      mod->setAttr(intel::TritonIntelGPUDialect::getLTSAttrName(),
+                   builder.getUnitAttr());
+
+    DPASAnalysis &dpasAnalysis = getAnalysis<DPASAnalysis>();
+    setThreadsPerWarp(mod, dpasAnalysis);
+  }
+
+private:
+  void setThreadsPerWarp(ModuleOp &mod,
+                         const DPASAnalysis &dpasAnalysis) const {
+    Builder builder(mod);
+    const std::string &AttrNumThreadsPerWarp =
+        TritonGPUDialect::getThreadsPerWarpAttrName();
+    unsigned reqThreadsPerWarp =
+        DPASAnalysis::supportedThreadsPerWarp(intel::getDeviceArch(mod));
+
+    mod.walk([&](FunctionOpInterface funcOp) {
+      if (dpasAnalysis.canUseDPAS(funcOp) == DPASAnalysis::Result::Maybe) {
+        // Set the threads per warp attribute to allow dot operation to be
+        // lowered to DPAS instructions.
+        mod->setAttr(AttrNumThreadsPerWarp,
+                     builder.getI32IntegerAttr(reqThreadsPerWarp));
+        assert(dpasAnalysis.canUseDPAS(funcOp) == DPASAnalysis::Result::True &&
+               "DPASAnalysis should report that dot operations can be "
+               "lowered to DPAS instructions");
+        WalkResult::interrupt();
```

Comment:
do we need to return this?
```suggestion
        return WalkResult::interrupt();
```

Reply:
We do not need to because when the walk is interrupted (hits that stmt), we would have set the threads-per-warps attribute on the module. Conversely, when the walk is not interrupted, we check the presence of that attribute in the module, because the attribute would not be found, we set it in the code at lines 64-67.

Reply:
An alternative implementation is possible. We could rely on the walk result to determine whether we need to set the threads-per-warp attribute on the module after the walk. But checking directly its presence on the module is prob. more robust.

Reply:
That statement simply creates a `WalkResult::Interrupt` instance. I think we need to return it to actually stop the walk.

Reply:
The documentation isn't clear. In upstream MLIR I see n mlir I see this case (./test/lib/Dialect/Vector/TestVectorTransforms.cpp):
```
    getOperation().walk([&](Operation *op) {
      if (auto warpOp = dyn_cast<WarpExecuteOnLane0Op>(op)) {
        if (hoistUniform) {
          moveScalarUniformCode(warpOp);
        }
        WalkResult::interrupt();
      }
    });
```
But ... better return it to be sure.

Reply:
fixed.